instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/HttpServerRequestMetrics.java

```diff
@@ -0,0 +1,132 @@
+
+  public static RequestMetricsFactory factory() {
+    return HttpServerRequestMetrics::new;
+  }
+
+  private final LongUpDownCounter activeRequests;
+  private final DoubleValueRecorder duration;
+
+  private HttpServerRequestMetrics(Meter meter) {
+    activeRequests =
+        meter
+            .longUpDownCounterBuilder("http.server.active_requests")
+            .setUnit("requests")
+            .setDescription("The number of concurrent HTTP requests that are currently in-flight")
+            .build();
+
+    duration =
+        meter
+            .doubleValueRecorderBuilder("http.server.duration")
+            .setUnit("milliseconds")
+            .setDescription("The duration of the inbound HTTP request")
+            .build();
+  }
+
+  @Override
+  public Context start(Context context, Attributes requestAttributes) {
+    long startTimeNanos = System.nanoTime();
+    Labels activeRequestLabels = activeRequestLabels(requestAttributes);
+    Labels durationLabels = durationLabels(requestAttributes);
+    activeRequests.add(1, activeRequestLabels);
+
+    return context.with(
+        HTTP_SERVER_REQUEST_METRICS_STATE,
+        new AutoValue_HttpServerRequestMetrics_State(
+            activeRequestLabels, durationLabels, startTimeNanos));
+  }
+
+  @Override
+  public void end(Context context, Attributes responseAttributes) {
+    State state = context.get(HTTP_SERVER_REQUEST_METRICS_STATE);
+    if (state == null) {
+      logger.debug(
+          "No state present when ending context {}. Cannot reset HTTP request metrics.", context);
+      return;
+    }
+    activeRequests.add(-1, state.activeRequestLabels());
+    duration.record(System.nanoTime() - state.startTimeNanos(), state.durationLabels());
+  }
+
+  private static Labels activeRequestLabels(Attributes attributes) {
+    LabelsBuilder labels = Labels.builder();
+    attributes.forEach(
+        (key, value) -> {
+          if (key.getType() != AttributeType.STRING) {
+            return;
+          }
+          switch (key.getKey()) {
+            case "http.method":
+            case "http.host":
+            case "http.scheme":
+            case "http.flavor":
+            case "http.server_name":
+              labels.put(key.getKey(), (String) value);
+              break;
+            default:
+              // fall through
+          }
+        });
+    return labels.build();
+  }
+
+  private static Labels durationLabels(Attributes attributes) {
+    LabelsBuilder labels = Labels.builder();
+    attributes.forEach(
+        (key, value) -> {
+          if (key.getType() != AttributeType.STRING) {
+            return;
+          }
+          switch (key.getKey()) {
+            case "http.method":
+            case "http.host":
+            case "http.scheme":
+            case "http.status_code":
+            case "http.flavor":
+            case "http.server_name":
+            case "net.host.name":
+            case "net.host.port":
```

Comment:
I'm curious about OTel's approach to high cardinality metric attributes, having implemented similar functionality in at Datadog and paid a lot of attention to cardinality reduction prior to aggregation. I have a few questions which I hope are not disruptive to your review.
* I'm unfamiliar with OTel attributes but which of these, if any, is the request path? 
* Since the unsanitised request path can have unbounded cardinality, would that rule it out as a metric attribute?
* If this were a metric for a database query, would you sanitise the query here, or exclude it from the labels?

Reply:
Thanks for reaching out @richardstartin. I doubt I'll have good answers as lots about OTel's approach to metrics is WIP but let's see ;)

> I'm unfamiliar with OTel attributes but which of these, if any, is the request path?

It's not in this list yet. There is a special note here

https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/semantic_conventions/http-metrics.md#parameterized-labels

and I didn't want to deal with it yet :) In particular, it seems they are not very up to date as `http.route` is really the parameter to be using (it's the parameterized path)

> Since the unsanitised request path can have unbounded cardinality, would that rule it out as a metric attribute?

I think so, it would have to be a parameterized path of some sort. That being said, given the separation between API and SDK in OTel, I wonder if it's possible to pass raw paths in API and in the SDK either detect parameters in the path using some cool logic, or detect the cardinality and drop the attributes. Without that, it's definitely not appropriate. I wonder if your "cardinality reduction prior to aggregation" would be something we can incorporate in the SDK.

> If this were a metric for a database query, would you sanitise the query here, or exclude it from the labels?

If there were conventions for DB metrics I'd just follow them

https://github.com/open-telemetry/opentelemetry-specification/tree/main/specification/metrics/semantic_conventions

Lacking them, just from my sense, I've never added queries as metric labels. But I don't see a problem with adding prepared statements as they generally have low cardinality (10-20 different types of queries in an app, each with the same prepared form across invocations).

Reply:
Thanks for your answers. We track metrics at a much more granular level than here - so including the sanitised request path, sanitised database query, and so on - we call this the _resource name_ at Datadog. This is because latency distributions tend to depend on attributes like the query or the endpoint - e.g. database updates and inserts tend to perform differently, and the application's business logic is always slower than the health check endpoint. 

I will be interested to see the direction this API goes in - will higher cardinality attributes be sacrificed altogether to keep the number of points down, or will it find a way to share the sanitisation required to avoid the number of metric points exploding with tracing to reduce total overhead and promote consistency between metrics and tracing?